State machine update

thedcap opened the pull request.

thedcap committed:
State Machine Rewrite
I used a method with classes
It seemed easier to organize that way. But I also cut down
on many of the states so this can be applied to Matthew's
design if we prefer it.

thedcap committed:
Filling in various classes of state machine
This commit has a lot of changes throughout the
state machine design. All of the process and
update functions are filled in save for the one
meant to perform the ensembles.

hgrehm commented on state_machine_design.py:
```diff
@@ -0,0 +1,355 @@
+
+import json
+import time
+import sys
```
Please sort imports (abc order)

hgrehm commented on state_machine_design.py:
```diff
@@ -0,0 +1,355 @@
+TIME_SHUTDOWN = 5 # find real shutdon and wakeup times later
+TIME_WAKEUP = 5
+
+def hmsToSeconds(hour: int, min: int, sec: int):
```
snake_case for method names

hgrehm commented on state_machine_design.py:
```diff
@@ -0,0 +1,355 @@
+        # transition to ITERATE
+        if self.check_time_ctrl == SKIP:
+            # sm.state = ITERATE() # switch with below line to disable error log on skip
+            sm.state = ERROR() # do we want to log this as error then recover?
```
I think using the error state partially as a logger could be good. For example, the missed ensemble error isn't worth stopping execution over but would be good to log. Might require some different logic for different errors though.

thedcap replied:
The only error right now that will shut down execution is the missing ensembles file. Right now if we skip an ensemble, the ERROR state should print out the error and then redirect control so that it then checks the next ensemble in the list.

hgrehm replied:
That sounds good for now then. I might address logging in a future commit since we haven't really been doing it all along, but printing is fine for the time being.

hgrehm commented on state_machine_design.py:
```diff
@@ -0,0 +1,355 @@
+                sm.err_code = MISSED_ENS
+                # sm.err_msg = "Current ensemble past time, skipping ensemble "
+                # + sm.ens_list[sm.ens_index]["title"] + " at " + curr_time_seconds
+            elif nearest_ens_time == curr_time_seconds:
```
We might want to leave more room for error here. Say, if `nearest_ens_time` hasn't passed yet but is within 5 seconds we could go ahead and execute.

thedcap replied:
Just so I'm understanding correctly, you're saying we can run an ensemble up to about 5 seconds _early_?

hgrehm replied:
Yes. We might play with the exact interval, but I think we'd rather execute slightly early than accidentally miss the single second when we should perform the ensemble.

hgrehm commented on state_machine_design.py:
```diff
@@ -0,0 +1,355 @@
+            sm.state = SLEEP()
+        # if all ensembles are done, need to sleep til first one of next day
+        elif self.check_time_ctrl == RESET:
+            sm.state = SLEEP()
```
I don't expect this to be something we use on deployment, but it'd be good to make this robust to the first ensemble needing to be performed immediately on reset. That is, if we've finished the last ensemble and `self.check_time_ctrl == RESET`, is there a case in which we would need to immediately transition to `PERFORM_ENSEMBLE` and can't afford the time to go to the `SLEEP` state?

thedcap replied:
I think that the only time that this could happen is if the final ensemble finishes at the exact time of, or after the start time of the very first ensemble. That seemed unlikely to me so I didn't account for it. I've already written changes for the other issues so I'll think over this one tomorrow and make it able to handle that situation if you think it's a good idea to be able to. I don't think it'll be too difficult.

hgrehm replied:
Ok, I don't think it's a big deal but if you have time to include it I'd appreciate it.

thedcap committed:
Added code for cases in moving from day to day
It should be able to handle waking up before/during/at the same time
as the first ensemble after handling the final ensemble of the
previous day.
Also some minor formatting changes (sorting input, renaming states,
proper function names)

hgrehm commented on state_machine_design.py:
```diff
@@ -118,30 +118,45 @@ def process(self, sm):
-                # + sm.ens_list[sm.ens_index]["title"] + " at " + curr_time_seconds
-            elif nearest_ens_time == curr_time_seconds:
+                print("time past current ens, checking if should skip")
+                if sm.rst == True:
+                    if now.tm_wday == sm.day_of_ens:
+                        self.check_time_ctrl = WAIT
+                    else:
+                        self.check_time = SKIP
+                        sm.err_code = MISSED_ENS
+                else:
+                    self.check_time_ctrl = SKIP
+                    sm.err_code = MISSED_ENS
```
Is there a difference between `self.check_time` and `self.check_time_ctrl`? I don't see the former anywhere else.
If they're the same, can we combine this into a single if-else statement, with the two conditions ANDed together, since the elses are the same?

thedcap replied:
Oop, apologies. I was writing late last night and typo'd that. Yeah I can fix that and combine them as well.

hgrehm commented on state_machine_design.py:
```diff
@@ -118,30 +118,45 @@ def process(self, sm):
-            elif nearest_ens_time == curr_time_seconds:
+                print("time past current ens, checking if should skip")
+                if sm.rst == True:
+                    if now.tm_wday == sm.day_of_ens:
```
I might be missing something, but under what scenario does `sm.day_of_ens` not equal `now.tm_wday`? From my understanding, this is trying to test whether the next ensemble is supposed to execute the day after the current one. How would `sm.day_of_ens` be set to reflect this?

thedcap replied:
This covers the situation where the final ensemble of day 1 begins before midnight but finishes after midnight. So the `sm.day_of_ens` would be day 1 and `now.tm_wday` would be day 2. The purpose is to differentiate finishing an ensemble during day 1 and the time being measured as later than the first ensemble of day 2, ie. finish at 10 pm and first ensemble of the day is 6 am (so the scheduler should go to sleep but the simple if statement says 10 pm is past 6 am and should be skipped) vs an ensemble finishing past the start time of the next ensemble, ie finish at 7 am and first ensemble of the day is 6 am (in which case the scheduler should skip the missed ensemble).

I think this scenario in actuality would be very rare (though I guess depends on when the ensembles are scheduled and how long they might actually take) but I tried to solve it just to have everything covered.

hgrehm replied:
Ah I see, thank you for explaining the approach.

thedcap committed:
Fix for mistyped variable name
